std/haxe/iterators/ArrayIterator.hx

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C)2005-2018 Haxe Foundation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package haxe.iterators;
+
+/**
+	This iterator is used only when `Array<T>` is passed to `Iterable<T>`
+**/
+class ArrayIterator<T> {
+	final array:Array<T>;
+	var current:Int = 0;
+
+	/**
+		Create a new `ArrayIterator`.
+	**/
+	public inline function new(array:Array<T>) {
+		this.array = array;
+	}
+
+	/**
+		See `Iterator.hasNext`
+	**/
+	public inline function hasNext() {
+		return current < array.length;
```

Comment:
maybe it would be a good time to decide if we want to cache the length or not

Reply:
It would be a breaking change.

Reply:
I don't see the point because this is undefined behavior anyway.

Reply:
Well, modification during iteration was always undefined and target-specific. Your change makes it consistent which is good, but is also breaking as it uncaches length for e.g. C# and Java, but this is also fine because, as said - this is officially undefined behaviour. I'd argue that we should cache the len for performance (less field access/getter calls), but maybe it's not worth it, don't know :-/

Reply:
Array iteration works without length caching for years (well, except C# and Java when typed as `Iteratable<T>`). It may be unspecified, but it's pretty consistent already even without this PR and that's how it works in existing code bases. Changing that definitely would break someone's projects for no good reason. I think performance gain would be insignificant if any.